packages/@ourworldindata/grapher/src/lineLegend/LineLegendHelpers.ts

```diff
@@ -0,0 +1,286 @@
+    const popCandidate = (): PlacedSeries | undefined => {
+        if (popCandidateFromSubset) {
+            const candidate = popCandidateFromSubset(remainingCandidates)
+            if (candidate) {
+                const index = remainingCandidates.indexOf(candidate)
+                remainingCandidates.splice(index, 1)
+            }
+            return candidate
+        }
+
+        return remainingCandidates.pop()
+    }
+
+    while (remainingCandidates.length > 0) {
+        const candidate = popCandidate()
+        if (!candidate) break
+
+        // sanity check if this is a valid candidate
+        if (!context.candidates.has(candidate)) continue
+
+        // either pick or dismiss the candidate
+        const newHeight = getNewHeight(context.keepSeriesHeight, candidate)
+        if (newHeight <= context.availableHeight) {
+            context = pickCandidate(context, candidate)
+            numPicked++
+        } else {
+            context = dismissCandidate(context, candidate)
+        }
+
+        // stop if we picked enough candidates
+        if (numPicked === maxCandidatesToPick) break
+    }
+
+    return context
+}
+
+/**
+ * Pick as many candidates as possible from a given subset.
+ *
+ * The order of candidates to consider for placement is determined by the
+ * `popCandidateFromSubset` function. The function should return the next candidate
+ * to consider. If the function returns `undefined`, the algorithm stops.
+ *
+ * If no custom function is provided, the algorithm picks candidates starting from
+ * the end (!) of the given list.
+ */
+export function pickAsManyAsPossibleWithRetry(
+    params: PickFromCandidateSubsetParams
+): LineLegendFilterAlgorithmContext {
+    return pickFromCandidateSubsetWithRetry(params)
+}
+
+/**
+ * Pick a fixed number of candidates from a give subset.
+ *
+ * The order of candidates to consider for placement is determined by the
+ * `popCandidateFromSubset` function. The function should return the next candidate
+ * to consider. If the function returns `undefined`, the algorithm stops.
+ *
+ * If no custom function is provided, the algorithm picks candidates starting from
+ * the end (!) of the given list.
+ */
+export function pickCandidateWithRetry(
+    params: PickFromCandidateSubsetParams
+): LineLegendFilterAlgorithmContext {
+    return pickFromCandidateSubsetWithRetry({
+        ...params,
+        maxCandidatesToPick: 1,
+    })
+}
+
+export function pickCandidateWithMaxDistanceToReferenceCandidate(params: {
+    context: LineLegendFilterAlgorithmContext
+    candidateSubset: PlacedSeries[]
+    referenceCandidate: PlacedSeries
+}): LineLegendFilterAlgorithmContext {
+    const { context, candidateSubset, referenceCandidate } = params
+    const distanceMap = new Map(
+        candidateSubset.map((c) => [c.seriesName, dist(c, referenceCandidate)])
+    )
+    const pop = (candidates: PlacedSeries[]) =>
+        maxBy(candidates, (c) => distanceMap.get(c.seriesName))
```

Comment:
We don't really need the intermediate step of creating the map here, I'd rather just do it like so:
```suggestion
    const pop = (candidates: PlacedSeries[]) =>
        maxBy(candidates, (c) => dist(c, referenceCandidate))
```